### PR TITLE
feature: non-finalized orders budget

### PR DESCRIPTION
**Pull request by mncdg**
This PR implements a budget (a hard cap) of all successfully fulfilled orders' value (expressed in USD) that were not reached
yet guaranteed finality at the given point in time.

For example, if you have allowed to fulfill orders worth $1 from this chain after 1 block confirmation:

```ts
      constraints: {
        requiredConfirmationsThresholds: [
           // worth <$100: 1+ block confirmation
           {thresholdAmountInUSD: 100, minBlockConfirmations: 1},
        ],
        // worth >$100: guaranteed block confirmations (12)
     }
```

and there is an accidental flood of 100,000 orders worth $1 occurs, you probably want to prevent this by setting the budget for non-finalized orders. If you set `nonFinalizedTVLBudget` to "100", than only first hundred of one-dollar orders would be attempted to be fulfilled, and all other orders would be postponed to the internal queue where they would be pulled one by one as soon as fulfilled orders are being finalized:

```ts
      constraints: {
        requiredConfirmationsThresholds: [
           // worth <$100: 1+ block confirmation
           {thresholdAmountInUSD: 100, minBlockConfirmations: 1},
        ],
        // worth >$100: guaranteed block confirmations (12)

        // Defines a TVL hard cap for orders coming from this chain that were fulfilled before getting guaranteedly finalized.
        nonFinalizedTVLBudget: 100,
     }
```

Task ID: DEV-2505

### CODE REVIEW

File: src/processors/universal.ts [hunk 1]
```diff
@@ -452,7 +458,8 @@ class UniversalProcessor extends BaseOrderProcessor {
       }
       else if ('Finalized' in finalizationInfo) {
         // do nothing: order have stable finality according to the WS
-        logger.debug('order source announced this order as finalized')
+        logger.debug('order source announced this order as finalized');
+        this.unconfirmedOrdersBudgetController.removeOrder(orderId, logger);
```
Comment (alexeychr) [inline src/processors/universal.ts:462]: additionally, add to the 'Fulfilled' event

File: src/processors/universal.ts [hunk 2]
```diff
@@ -737,6 +747,10 @@ while calculateExpectedTakeAmount returned ${tokenAddressToString(orderInfo.orde
       return;
     }
 
+    if (orderIsConfirmed) {
+      this.unconfirmedOrdersBudgetController.validateAndAddOrder(orderId, usdWorth, logger);
+    }
+
```
Comment (alexeychr) [inline src/processors/universal.ts:753]: This is nonsense: you literally add confirmed order (isConfirmed==true) to the list of unconfirmed orders. How can this happen? 

Also, why do you perform validation after the order has been fulfilled, why?

File: src/processors/universal.ts [hunk 3]
```diff
@@ -709,6 +716,9 @@ while calculateExpectedTakeAmount returned ${tokenAddressToString(orderInfo.orde
       (fulfillTx as Tx).cappedGasPrice = evmFulfillCappedGasPrice;
     }
 
+    if (orderIsConfirmed) {
+      this.unconfirmedOrdersBudgetController.validateOrder(orderId, usdWorth, logger);
```
Comment (alexeychr) [inline src/processors/universal.ts:720]: Perform validation in the `'Confirmed' in finalizationInfo` section

File: src/processors/UnconfirmedOrdersBudgetController.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    constructor(private readonly unconfirmedOrdersBudgetInUSD?: number) {}
+
+    validateOrder(orderId: string, usdWorth: number, orderLogger: Logger): boolean|never {
+        if (!this.unconfirmedOrdersBudgetInUSD) {
+            return true;
+        }
+        const logger = orderLogger.child({ service: UnconfirmedOrdersBudgetController.name });
+        const potentialSpentBudgetInUSD = this.spentBudgetInUSD + usdWorth - (this.orders.get(orderId) || 0);
+        if (potentialSpentBudgetInUSD > this.unconfirmedOrdersBudgetInUSD) {
+            const message = `Order with usd worth ${usdWorth} is out of budget ${this.unconfirmedOrdersBudgetInUSD}(spent ${this.spentBudgetInUSD})`;
```
Comment (alexeychr) [inline src/processors/UnconfirmedOrdersBudgetController.ts:17]: misleading message

**Review by alexeychr — CHANGES_REQUESTED** [review]
Have breaches in the logic flow.

Add description

**Review by alexeychr — CHANGES_REQUESTED** [review]
Have breaches in the logic flow.

Add description

**Comment by merrymari** [comment]
Task linked: [DEV-2505 В dln taker в настройках custom finality сделать потолок суммы долларовых эквивалентов всех не финализированных ордеров](https://app.clickup.com/t/4717986/DEV-2505)

**Review by alexeychr — APPROVED** [review]

**Pull request merged**